### DISCUSSION TITLE
[pom editor question] How does the pom editor resolve the local repository to analyze dependencies?

### DISCUSSION THREAD

**Discussion by reto-hoehener** [eclipse-m2e/m2e-core#1809] [Q&A]
I ran maven in the cmd prompt once with a custom local repo parameter (`-Dmaven.repo.local=reactor-repo`), and now the m2e pom editor also uses that repo for dependency analysis and navigation. I get many warnings that plugins cannot be found.

Sometimes I use a custom user settings file in the maven settings to point to (yet another) custom local repo, but currently I have no special configuration (see screenshot).

I tried search for 'reactor-repo' in the eclipse workspace and plugin folders but didn't find anything.

Could you help me debug how m2e resolves the local repository?

![image](https://github.com/user-attachments/assets/44409dbf-9f30-44b9-995d-dac31e0e8dc8)

![image](https://github.com/user-attachments/assets/7adaa03d-bce2-4bb9-a43b-8d7bfb66d3a0)

**Comment by laeubi** [comment]
Have you tried "update maven project" with the force option?

**Reply by reto-hoehener** [reply]
Yes, this is always the first thing I do: Restart eclipse, then select all projects and update maven projects with force update snapshots.

**Comment by reto-hoehener** [comment]
I have also tried deleting all `.lemminx-maven` folders I could find. It looks to me as if m2e keeps it's own repository there? It looks like it creates that folder next to wherever the 'currently identified local repository' is.

![image](https://github.com/user-attachments/assets/5c3cc24c-a86e-45ce-a535-0f0cca718e6c)

**Comment by reto-hoehener** [comment]
I tried to reproduce it with a simple new workspace but no luck so far. But all my real big workspaces are messed up as soon as I start switching between custom user settings and different local repos.

It would already help a lot if I could find the place where m2e stores the location of the local repository (I assume it is some kind of stale cache).

**Comment by mickaelistria** [comment]
The LemMinX Language Server and its extension for Maven get informed of the path to settings file configured via preferences during their initialization, see https://github.com/eclipse-m2e/m2e-core/blob/a1dad4438eba4d895713a7cd039f51e04472d730/org.eclipse.m2e.editor.lemminx/src/org/eclipse/m2e/editor/lemminx/InitializationOptionsProvider.java#L29 .
The LemMinX extension is supposed to receive it and process it, in [org.eclipse.lemminx.extensions.maven.MavenLemminxExtension.initMavenRequest()](https://github.com/eclipse/lemminx-maven/blob/d0857b05f9ed858780c578c6b661ab84f3628d4d/lemminx-maven/src/main/java/org/eclipse/lemminx/extensions/maven/MavenLemminxExtension.java#L392) .

If you want to debug the running language server, please see https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/blob/master/TIPS_and_FAQ.md#attach-a-debugger-to-the-xml-language-server-process

**Reply by reto-hoehener** [reply]
Thank you Mickael, I will have a look!

**Comment by reto-hoehener** [comment]
I figured out the reason for the "plugin could not be resolved" warning. The cause is given as "Failed to parse plugin descriptor for ...", followed by a workspace project path (not a local repo path). So I guess the POM editor tries to use workspace resolution to read `<plugin-project>/src/main/resources/META-INF/maven/plugin.xml` (or the corresponding location in the target dir), which does not exist (generated during packaging and only exists in the final plugin jar).

![image](https://github.com/user-attachments/assets/8d230263-828c-47fa-ab2c-3fa3166c3a12)

My current workarounds are:
- close the plugin project -> reopen the pom.xml -> pom editor resolves correctly from local repo (no warnings, goal autocompletion works)
- or bump the plugin-project-version to something different than the plugin-version referenced in all the other workspace projects
